src/linux/native/jgroup/com/alibaba/tenant/jgroup.c

```diff
@@ -0,0 +1,402 @@
+            }                                   \
+        } while(0)
+
+// Environment variable name to turn on debugging output of JGroup native code
+#define JGROUP_DEBUG_ENV_KEY    "JGROUP_DEBUG"
+
+// logging support
+#define JGROUP_LOG(stream, format, ...)  do {           \
+            if (jgroup_debug == JNI_TRUE) {             \
+                fprintf(stream, format, __VA_ARGS__);   \
+            }                                           \
+        } while (0)
+
+#define JGROUP_MSG(msg)             JGROUP_LOG(stdout, "[JGROUP][INFO] %s\n", msg)
+#define JGROUP_INFO(format, ...)    JGROUP_LOG(stdout, "[JGROUP][INFO]" format "\n", __VA_ARGS__)
+#define JGROUP_WARN(format, ...)    JGROUP_LOG(stderr, "[JGROUP][WARN]" format "\n", __VA_ARGS__)
+#define JGROUP_ERROR(format, ...)   JGROUP_LOG(stderr, "[JGROUP][ERROR]" format "\n", __VA_ARGS__)
+
+#define SIG_STRING              "Ljava/lang/String;"
+#define MAX_PATH_LEN            2048
+#define CG_MAX_CTRL_NAME_LEN    128
+
+//--------------------- Utility methods ----------------------
+
+// Returns thread ID of current thread
+static pid_t get_tid() {
+    return syscall(SYS_gettid);
+}
+
+// returns JNI_TRUE if a 'path' is a file (not a directory); otherwise JNI_FALSE
+static jboolean
+is_file(const char* path) {
+    if (path != NULL) {
+        struct stat st;
+        memset(&st, 0, sizeof(struct stat));
+        if (0 == stat(path, &st)) {
+            return S_ISREG(st.st_mode) ? JNI_TRUE : JNI_FALSE;
+        }
+    }
+    return JNI_FALSE;
+}
+
+//--------------------- JNI impl ----------------------
+
+// Detect if cgroup is enabled by reading /proc/cgroups
+static int
+detect_cgroup_enabled() {
+    char subsys_name[CG_MAX_CTRL_NAME_LEN];
+    int res = 0, err = 0;
+    int hierarchy = 0, num_cgroups = 0, enabled = 0;
+    const char* proc_cgroups_path = "/proc/cgroups";
+    int cg_index = 0;
+    FILE* proc_cgroups = fopen(proc_cgroups_path, "re");
+
+    if (proc_cgroups != NULL) {
+        while (!feof(proc_cgroups)) {
+            err = fscanf(proc_cgroups, "%s %d %d %d", subsys_name,
+                    &hierarchy, &num_cgroups, &enabled);
+            if (err < 0) {
+                break;
+            }
+
+            for (cg_index = 0; cg_index < CG_MAX_SUPPORTED; ++cg_index) {
+                if (!strcmp(subsys_name, cg_controller_names[cg_index]) && enabled) {
+                    ENABLE_CONTROLLER(cg_index);
+                    JGROUP_INFO("cgroup controller %s is enabled!", cg_controller_names[cg_index]);
+                    break;
+                }
+            }
+        }
+        JGROUP_INFO("Opening %s to detect enabled controllers", proc_cgroups_path);
+    } else {
+        JGROUP_ERROR("Failed to open %s", proc_cgroups_path);
+        res = -1;
+    }
+
+    return res;
+}
+
+// Detect cgroup mountpoints
+static int
+detect_cgroup_mount_points() {
+    int res = 0;
+	struct mntent *ent = NULL;
+	char mntent_buffer[4 * MAX_PATH_LEN];
```

Comment:
alignment

Reply:
done.
there were mixed tabs and spaces, just removed tabs.